codec-redis/src/main/java/io/netty/handler/codec/redis/RedisDecoder.java

```diff
@@ -0,0 +1,298 @@
+        if (in.readableBytes() < RedisConstants.EOL_LENGTH) {
+            return false;
+        }
+        readEndOfLine(in);
+        out.add(FullBulkStringRedisMessage.EMPTY_INSTANCE);
+        resetDecoder();
+        return true;
+    }
+
+    // ${expectedBulkLength}\r\n <here> {data...}\r\n
+    private boolean decodeBulkStringContent(ByteBuf in, List<Object> out) throws Exception {
+        final int readableBytes = in.readableBytes();
+        if (readableBytes == 0) {
+            return false;
+        }
+
+        // if this is last frame.
+        if (readableBytes >= remainingBulkLength + RedisConstants.EOL_LENGTH) {
+            ByteBuf content = in.readSlice(remainingBulkLength).retain();
+            readEndOfLine(in);
+            out.add(new DefaultLastBulkStringRedisContent(content));
+            resetDecoder();
+            return true;
+        }
+
+        // chunked write.
+        int toRead = Math.min(remainingBulkLength, readableBytes);
+        remainingBulkLength -= toRead;
+        out.add(new DefaultBulkStringRedisContent(in.readSlice(toRead).retain()));
+        return true;
+    }
+
+    private static void readEndOfLine(final ByteBuf in) {
+        final short delim = in.readShort();
+        if (RedisConstants.EOL_SHORT == delim) {
+            return;
+        }
+        final byte[] bytes = RedisCodecUtil.shortToBytes(delim);
+        throw new RedisCodecException("delimiter: [" + bytes[0] + "," + bytes[1] + "] (expected: \\r\\n)");
+    }
+
+    private RedisMessage newInlineRedisMessage(RedisMessageType messageType, ByteBuf content) {
+        switch (messageType) {
+        case SIMPLE_STRING: {
+            SimpleStringRedisMessage cached = messagePool.getSimpleString(content);
+            return cached != null ? cached : new SimpleStringRedisMessage(content.toString(CharsetUtil.UTF_8));
+        }
+        case ERROR: {
+            ErrorRedisMessage cached = messagePool.getError(content);
+            return cached != null ? cached : new ErrorRedisMessage(content.toString(CharsetUtil.UTF_8));
+        }
+        case INTEGER: {
+            IntegerRedisMessage cached = messagePool.getInteger(content);
+            return cached != null ? cached : new IntegerRedisMessage(parseRedisNumber(content));
+        }
+        default:
+            throw new RedisCodecException("bad type: " + messageType);
+        }
+    }
+
+    private static ByteBuf readLine(ByteBuf in) {
+        final int lfIndex = in.forEachByte(ByteProcessor.FIND_LF);
+        if (lfIndex < 0) {
+            return null;
+        }
+        ByteBuf data = in.readSlice(lfIndex - in.readerIndex() - 1); // `-1` is for CR
+        readEndOfLine(in); // validate CR LF
+        return data;
+    }
+
+    private long parseRedisNumber(ByteBuf byteBuf) {
+        final boolean negative = byteBuf.getByte(byteBuf.readerIndex()) == '-';
+        if (byteBuf.readableBytes() > RedisConstants.POSITIVE_LONG_MAX_LENGTH + (negative ? 1 : 0)) {
```

Comment:
should we also validate the case where there are not enough characters to parse a number (feel free to restructure / optimize conditional checks):

```java
int readableBytes = byteBuf.readableBytes();
if (!negative && readableBytes == 0 || negative && readableBytes == 1) {
  // there is no number to parse! we shouldn't return 0 but instead throw an exception
}
```